python/datafusion/expr.py

```diff
@@ -15,9 +15,256 @@
+
+    def cast(self, to: pa.DataType[Any]) -> Expr:
+        return Expr(self.expr.cast(to))
+
+    def rex_type(self) -> RexType:
+        return self.expr.rex_type()
+
+    def types(self) -> DataTypeMap:
+        return self.expr.types()
+
+    def python_value(self) -> Any:
+        return self.expr.python_value()
+
+    def rex_call_operands(self) -> list[Expr]:
+        return [Expr(e) for e in self.expr.rex_call_operands()]
+
+    def rex_call_operator(self) -> str:
+        return self.expr.rex_call_operator()
+
+    def column_name(self, plan: LogicalPlan) -> str:
+        return self.expr.column_name()
+
+
+class WindowFrame:
+    def __init__(
+        self, units: str, start_bound: int | None, end_bound: int | None
+    ) -> None:
+        """
+        :param units: Should be one of `rows`, `range`, or `groups`
+        :param start_bound: Sets the preceeding bound. Must be >= 0. If none, this will be set to unbounded. If unit type is `groups`, this parameter must be set.
+        :param end_bound: Sets the following bound. Must be >= 0. If none, this will be set to unbounded. If unit type is `groups`, this parameter must be set.
+        """
+        self.window_frame = expr_internal.WindowFrame(units, start_bound, end_bound)
+
+    def get_frame_units(self) -> str:
+        """
+        Returns the window frame units for the bounds
+        """
+        return self.window_frame.get_frame_units()
+
+    def get_lower_bound(self) -> WindowFrameBound:
+        """
+        Returns starting bound
+        """
+        return WindowFrameBound(self.window_frame.get_lower_bound())
+
+    def get_upper_bound(self):
+        """
+        Returns end bound
+        """
+        return WindowFrameBound(self.window_frame.get_upper_bound())
+
+
+class WindowFrameBound:
+    def __init__(self, frame_bound: expr_internal.WindowFrameBound) -> None:
+        self.frame_bound = frame_bound
+
+    def get_offset(self) -> int | None:
+        """
+        Returns the offset of the window frame
+        """
+        return self.frame_bound.get_offset()
+
+    def is_current_row(self) -> bool:
+        """
+        Returns if the frame bound is current row
+        """
+        return self.frame_bound.is_current_row()
+
+    def is_following(self) -> bool:
+        """
+        Returns if the frame bound is following
+        """
+        return self.frame_bound.is_following()
+
+    def is_preceding(self) -> bool:
+        """
+        Returns if the frame bound is preceding
+        """
+        return self.frame_bound.is_preceding()
+
+    def is_unbounded(self) -> bool:
+        """
+        Returns if the frame bound is unbounded
+        """
+        return self.frame_bound.is_unbounded()
+
+
+class CaseBuilder:
+    def __init__(self, case_builder: expr_internal.CaseBuilder) -> None:
+        """
+        :param case_builder: Internal object. This constructor is not expected to be used by the end user. Instead use :func:`case` to construct.
```

Comment:
Perhaps the docstring style should be consistent with the other modules?

Reply:
Good point. Do you have a preferred style? I have no strong opinions on which style as long as it's consistently applied.

Reply:
Don't have an opinion on a specific style but it looks like ruff can be configured to enforce a specific style, which is probably worth enabling https://docs.astral.sh/ruff/settings/#lint_pydocstyle_convention

Reply:
Likewise, I think just enforcing some consistent style (regardless of what it is, I usually stick to the Google style guide) would be good.

Reply:
Works for me. Thanks! I'll work on it.

Reply:
I've added the google style to the ruff stage. It caught a lot of things, but didn't catch these mixed styles. I'll try to look further tomorrow, but did do a massive update on getting the whitespace and flushing out those with mixing docstrings.